contracts/child/ChildToken/ChildMintableERC721.sol

```diff
@@ -0,0 +1,92 @@
+pragma solidity ^0.6.6;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {IChildToken} from "./IChildToken.sol";
+import {NetworkAgnostic} from "../../common/NetworkAgnostic.sol";
+import {ChainConstants} from "../../ChainConstants.sol";
+
+
+contract ChildMintableERC721 is ERC721, IChildToken, AccessControl, NetworkAgnostic, ChainConstants {
+    bytes32 public constant DEPOSITOR_ROLE = keccak256("DEPOSITOR_ROLE");
+    mapping (uint256 => bool) public withdrawnTokens;
+
+    constructor(
+        string memory name_,
+        string memory symbol_
+    ) public ERC721(name_, symbol_) NetworkAgnostic(name_, ERC712_VERSION, ROOT_CHAIN_ID) {
+        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+        _setupRole(DEPOSITOR_ROLE, _msgSender());
+    }
+
+    modifier only(bytes32 role) {
+        require(hasRole(role, _msgSender()), "ChildMintableERC721: INSUFFICIENT_PERMISSIONS");
+        _;
+    }
+
+    function _msgSender()
+        internal
+        override
+        view
+        returns (address payable sender)
+    {
+        if (msg.sender == address(this)) {
+            bytes memory array = msg.data;
+            uint256 index = msg.data.length;
+            assembly {
+                // Load the 32 bytes word from memory with the address on the lower 20 bytes, and mask those.
+                sender := and(
+                    mload(add(array, index)),
+                    0xffffffffffffffffffffffffffffffffffffffff
+                )
+            }
+        } else {
+            sender = msg.sender;
+        }
+        return sender;
+    }
+
+    /**
+     * @notice called when token is deposited on root chain
+     * @dev Should be callable only by ChildChainManager
+     * Should handle deposit by minting the required tokenId for user
+     * Should set `withdrawnTokens` mapping to `false` for the tokenId being deposited
+     * Minting can also be done by other functions
+     * @param user user address for whom deposit is being done
+     * @param depositData abi encoded tokenId
+     */
+    function deposit(address user, bytes calldata depositData)
+        external
+        override
+        only(DEPOSITOR_ROLE)
+    {
+        uint256 tokenId = abi.decode(depositData, (uint256));
+        withdrawnTokens[tokenId] = false;
+        _mint(user, tokenId);
+    }
+
+    /**
+     * @notice called when user wants to withdraw token back to root chain
+     * @dev Should handle withraw by burning user's token.
+     * Should set `withdrawnTokens` mapping to `true` for the tokenId being withdrawn
+     * This transaction will be verified when exiting on root chain
+     * @param tokenId tokenId to withdraw
+     */
+    function withdraw(uint256 tokenId) external {
```

Comment:
Should we add check for `owner` for `tokenId` just to make sure people don't directly deploy it?

Reply:
`withdraw` function checks if owner of `tokenId` is `_msgSender()`. Are you referring to some additional check?